src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,997 @@
+
+        if sweeper.parameter.name.lower() == "frequency":
+            for pulse in sweeper.pulses:
+                qubit = pulse.qubit
+                if pulse.type.name.lower() == "drive":
+                    line = "drive"
+                elif pulse.type.name.lower() == "readout":
+                    line = "measure"
+                exp_calib[f"{line}{qubit}"] = lo.SignalCalibration(
+                    oscillator=lo.Oscillator(
+                        frequency=ZhSweeper(pulse, sweeper, qubits[sweeper.pulses[0].qubit]).zhsweeper,
+                        modulation_type=lo.ModulationType.HARDWARE,
+                    )
+                )
+
+        if sweeper.parameter.name == "bias":
+            for qubit in sweeper.qubits.values():
+                # for qubit in sweeper.qubits.values():
+                #     qubit = sweeper.qubits[qubit]
+
+                parameter = ZhSweeperLine(sweeper, qubit, self.sequence_qibo).zhsweeper
+
+        if sweeper.parameter.name == "delay":
+            pulse = sweeper.pulses[0]
+            qubit = pulse.qubit
+            # for qubit in sweeper.qubits.values():
+            #     qubit = sweeper.qubits[qubit]
+
+            parameter = ZhSweeperLine(sweeper, qubit, self.sequence_qibo).zhsweeper
+
+        if parameter is None:
+            parameter = ZhSweeper(sweeper.pulses[0], sweeper, qubits[sweeper.pulses[0].qubit]).zhsweeper
+
+        with exp.sweep(
+            uid=f"sweep_{sweeper.parameter.name.lower()}_{i}",
+            parameter=parameter,
+            reset_oscillator_phase=True,
+        ):
+            if len(self.sweepers) > 0:
+                self.sweep_recursion(qubits, exp, exp_calib, relaxation_time, acquisition_type, fast_reset)
+            else:
+                self.select_exp(exp, qubits, relaxation_time, acquisition_type, fast_reset)
+
+    # -----------------------------------------------------------------------------
+
+    # TODO: Implement further pulse viewing functions from 2.2.0
+    # should this be added in a way so the user can check how the sequence looks like ?
+    def run_sim(self, sim_time):
+        self.device_setup = lo.DeviceSetup.from_descriptor(
+            yaml_text=self.descriptor,
+            server_host="localhost",
+            server_port="8004",
+            setup_name=self.name,
+        )
+        # create a session
+        self.sim_session = lo.Session(self.device_setup)
+        # connect to session
+        self.sim_device = self.session.connect(do_emulation=True)
+
+        # Plot simulated output signals with helper function
+        plot_simulation(
+            self.exp,
+            start_time=0,
+            length=sim_time,
+            plot_width=10,
+            plot_height=3,
+        )
+
+
+import re
+
+# additional imports for plotting
+import matplotlib.pyplot as plt
+import matplotlib.style as style
+
+# numpy for mathematics
+import numpy as np
+from matplotlib import cycler
```

Comment:
This is probably temporary here until we do it properly somewhere else, for all drivers, but if you plan to leave it here for now I would suggest to move the imports on top.

Alternatively you could also move the plot related imports inside the function that is using them (if it is only one function), so that they are not imported if not needed.

Reply:
Is there as I read something about improving plotting of the pulses on the next version. I still haven't explored that as much as it is not that important for the driver. So I am expecting that part to change. I will move the imports inside the function for now.